File: Python/trex_selector_methods/trex/demo_trex_07_mc_sim_mmap/demo_trex_07_mc_sim_mmap.py
# ==============================================================================
# demo_trex_07_mc_sim_mmap.py
# ==============================================================================
#
# Classical T-Rex selector demo — Monte Carlo simulations with memory-mapped
# workflows.
# Mirrors R/trex_selector_methods/trex/demo_trex_07_mc_sim_mmap.R and
# cpp/trex_selector_methods/trex/demo_trex_07_mc_sim_mmap.cpp.
#
# Demo content:
#
#  Part C: MC — In-memory X + use_memory_mapping=True.
#          Activates D-mmap and solver serialization inside TRexSelector.
#          Purpose: verify that the D-mmap + solver-serialization pipeline
#          produces stable, reproducible FDR/TPR statistics over many runs.
#          Uses run_mc_trex() (existing in-memory worker).
#
#  Part D: MC — Fully memory-mapped X + use_memory_mapping=True.
#          Each MC trial writes X to a unique temporary binary file, passes
#          X_mmap.to_numpy() to TRexSelector, and deletes the file in a
#          finally block — exception-safe RAII equivalent.
#          Uses run_mc_trex_full_mmap() (new full-mmap worker).
#
# Both parts should produce statistically equivalent FDR/TPR results since the
# data and algorithm are identical — the equivalence is the verification goal.
#
# Shared configuration (mirrors C++ Demo C/D and R Part C/D):
#   n=300, p=1000, s=10  (high-dim)
#   TLARS only
#   Fixed support drawn once with seed=24 (shared across all MC trials)
#   SNR sweep: {0.1, 0.5, 1.0, 2.0, 5.0}
#   tloop_max_stagnant_steps=7,  opt_threshold=0.75
#   num_MC=500
#
# Seed spaces (no clashes):
#   DGP (X, noise)   : trial_seed = base_seed + mc - 1
#   Fixed support     : make_support_random(seed=24) → internal seed+500_000
#   TRexSelector      : seed=-1 (hardware entropy, required for valid MC FDR)
#   Temp mmap paths   : tempfile.mkstemp() — OS-guaranteed unique per worker
#
# ==============================================================================

import sys
import os

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

import numpy as np

from support_generators import make_support_random
from trex_sim_common import (
    run_mc_trex,
    run_mc_trex_full_mmap,
    save_mc_results,
)

# ==============================================================================
# Global parameters
# ==============================================================================

_OUT_DIR     = os.path.join(_THIS_DIR, "simulation_results", "data")
_NUM_WORKERS = 6
_NUM_MC      = 500

# TLARS solver descriptor (only solver used in mmap demos)
_TLARS_SOLVER = dict(
    solver_name   = "TLARS",
    name          = "TLARS",
    lambda2       = 0.1,
    rho_afs       = 0.3,
    ncgmp_variant = 0,
)

# Shared base control for both Part C and Part D.
# use_memory_mapping=True: enables D-mmap + solver serialization (Part C's
# primary mode, and the "internal mmap layer" on top of X-mmap in Part D).
_MMAP_BASE_CTRL = dict(
    K                        = 20,
    max_dummy_multiplier     = 10,
    use_max_T_stop           = True,
    lloop_strategy           = "HCONCAT",
    tloop_stagnation_stop    = True,
    tloop_max_stagnant_steps = 7,
    opt_threshold            = 0.75,
    use_memory_mapping       = True,    # D-mmap + solver serialization
    parallel_rnd_experiments = False,   # mirrors R use_openmp=FALSE
)


# ==============================================================================
# Part C: MC — In-memory X + use_memory_mapping=True
# ==============================================================================

def trx_03_part_c(num_MC=_NUM_MC, num_workers=_NUM_WORKERS, high_dim=True):
    """
    Mirrors C++ demo_TRexSelector_d_mmap_MonteCarlo() and R trx_03_part_c().

    Uses run_mc_trex() (existing in-memory worker) with use_memory_mapping=True
    in the control dict — the C++ backend memory-maps the dummy matrices D and
    serializes solver checkpoints, but X stays in RAM.
    """
    n = 300  if high_dim else 1000
    p = 1000 if high_dim else 300
    s = 10
    tFDR       = 0.1
    snr_values = [0.1, 0.5, 1.0, 2.0, 5.0]

    dim_label = "High-dimensional (p > n)" if high_dim else "Low-dimensional (n > p)"
    print(f"\n{'=' * 70}")
    print(f"Part C: MC \u2014 In-Memory X + use_memory_mapping=True")
    print(f"  Verifies D-mmap + solver-serialization pipeline over many runs.")
    print(f"  {dim_label}  (n = {n}, p = {p},  num_MC = {num_MC})\n")

    # Fixed support shared across all MC runs — mirrors C++ rng(24) and R seed=24
    true_support = make_support_random(s, p, seed=24)
    print(f"True support (cardinality {s}): {{{', '.join(str(x) for x in true_support)}}}\n")

    n_snr     = len(snr_values)
    fdr_vec   = np.zeros(n_snr)
    tpr_vec   = np.zeros(n_snr)
    avg_L_vec = np.zeros(n_snr)
    avg_T_vec = np.zeros(n_snr)

    for snr_idx, snr in enumerate(snr_values):
        base_seed = 24 + snr_idx * 1000
        lbl = f"SNR={snr:.2f} [TLARS/mmap-D]"
        result = run_mc_trex(
            n=n, p=p, s=s, snr=snr,
            solver_desc=_TLARS_SOLVER,
            base_ctrl_dict=_MMAP_BASE_CTRL,
            tFDR=tFDR,
            base_seed=base_seed,
            num_MC=num_MC,
            max_workers=num_workers,
            fixed_support=true_support,
            track_L_T=True,
            label=lbl,
        )
        fdr_vec[snr_idx]   = result["mean_FDR"]
        tpr_vec[snr_idx]   = result["mean_TPR"]
        avg_L_vec[snr_idx] = result["mean_L"]
        avg_T_vec[snr_idx] = result["mean_T"]

    # Wrap into 1-row matrices to reuse save_mc_results()
    stagnant_sw = _MMAP_BASE_CTRL["tloop_max_stagnant_steps"]
    stem = f"d03_trex_mmap_demo_c_n{n}_p{p}_sw{stagnant_sw}"
    save_mc_results(
        _OUT_DIR, stem, num_MC,
        row_names=["TLARS"],
        snr_values=snr_values,
        fdr_mat=fdr_vec.reshape(1, -1),
        tpr_mat=tpr_vec.reshape(1, -1),
        avg_L_mat=avg_L_vec.reshape(1, -1),
        avg_T_mat=avg_T_vec.reshape(1, -1),
    )
    print()


# ==============================================================================
# Part D: MC — Fully memory-mapped X + use_memory_mapping=True
# ==============================================================================

def trx_03_part_d(num_MC=_NUM_MC, num_workers=_NUM_WORKERS, high_dim=True):
    """
    Mirrors C++ demo_TRexSelector_full_mmap_MonteCarlo() and R trx_03_part_d().

    Uses run_mc_trex_full_mmap() — each worker writes X to a unique per-trial
    temporary binary file via numpy_to_memmap(), passes X_mmap.to_numpy() to
    TRexSelector, and deletes the temp file in a finally block.

    use_memory_mapping=True in the control dict additionally activates D-mmap
    and solver serialization inside TRexSelector on top of the X-mmap layer.
    This is the full mmap pipeline: X mmap + D mmap + solver serialization.
    """
    n = 300  if high_dim else 1000
    p = 1000 if high_dim else 300
    s = 10
    tFDR       = 0.1
    snr_values = [0.1, 0.5, 1.0, 2.0, 5.0]

    dim_label = "High-dimensional (p > n)" if high_dim else "Low-dimensional (n > p)"
    print(f"\n{'=' * 70}")
    print(f"Part D: MC \u2014 Fully Memory-Mapped X + use_memory_mapping=True")
    print(f"  Full mmap pipeline: X mmap + D mmap + solver serialization.")
    print(f"  {dim_label}  (n = {n}, p = {p},  num_MC = {num_MC})\n")

    # Fixed support — same as Part C, enabling direct comparison
    true_support = make_support_random(s, p, seed=24)
    print(f"True support (cardinality {s}): {{{', '.join(str(x) for x in true_support)}}}\n")

    n_snr     = len(snr_values)
    fdr_vec   = np.zeros(n_snr)
    tpr_vec   = np.zeros(n_snr)
    avg_L_vec = np.zeros(n_snr)
    avg_T_vec = np.zeros(n_snr)

    for snr_idx, snr in enumerate(snr_values):
        base_seed = 24 + snr_idx * 1000
        lbl = f"SNR={snr:.2f} [TLARS/mmap-XD]"
        result = run_mc_trex_full_mmap(
            n=n, p=p, s=s, snr=snr,
            solver_desc=_TLARS_SOLVER,
            base_ctrl_dict=_MMAP_BASE_CTRL,
            tFDR=tFDR,
            base_seed=base_seed,
            num_MC=num_MC,
            max_workers=num_workers,
            fixed_support=true_support,
            track_L_T=True,
            label=lbl,
        )
        fdr_vec[snr_idx]   = result["mean_FDR"]
        tpr_vec[snr_idx]   = result["mean_TPR"]
        avg_L_vec[snr_idx] = result["mean_L"]
        avg_T_vec[snr_idx] = result["mean_T"]

    stagnant_sw = _MMAP_BASE_CTRL["tloop_max_stagnant_steps"]
    stem = f"d03_trex_mmap_demo_d_n{n}_p{p}_sw{stagnant_sw}"
    save_mc_results(
        _OUT_DIR, stem, num_MC,
        row_names=["TLARS"],
        snr_values=snr_values,
        fdr_mat=fdr_vec.reshape(1, -1),
        tpr_mat=tpr_vec.reshape(1, -1),
        avg_L_mat=avg_L_vec.reshape(1, -1),
        avg_T_mat=avg_T_vec.reshape(1, -1),
    )
    print()


# ==============================================================================
# Main
# ==============================================================================

if __name__ == "__main__":
    trx_03_part_c()
    trx_03_part_d()
    print("\ndemo_trex_07_mc_sim_mmap.py complete.")

File: cpp/trex_selector_methods/trex/demo_trex_07_mc_sim_mmap/demo_trex_07_mc_sim_mmap.cpp
// ==============================================================================
// demo_trex_07_mc_sim_mmap.cpp
// ==============================================================================
/**
 * @file demo_trex_07_mc_sim_mmap.cpp
 *
 * @brief Monte Carlo simulations for T-Rex Selector with memory-mapped workflows.
 *
 * @details Contains two Monte Carlo simulation scenarios that mirror the
 *          corresponding single-run variants in demo_trex_06_mmap.cpp:
 *
 *  (C) MC with in-memory X + `use_memory_mapping=true`:
 *      Activates solver serialization and memory-mapped dummy matrices D.
 *      Purpose: verify that the D-mmap + solver-serialization pipeline produces
 *      stable, reproducible FDR/TPR statistics over many runs.
 *
 *  (D) MC with fully memory-mapped pipeline (X + D + solver serialization):
 *      Each MC iteration creates its own mmap-backed X/y files and destroys them
 *      via a RAII guard at the end of the iteration scope — exception-safe.
 *
 * Results are written as:
 *  - Aligned text table (.txt) to simulation_results/ (dual console + file output)
 *  - Tidy long-format CSV (.csv) to simulation_results/ for plotting
 */
// ==============================================================================

// std includes
#include <iomanip>
#include <iostream>
#include <map>
#include <random>
#include <string>
#include <unordered_set>
#include <vector>

// Eigen includes
#include <Eigen/Dense>

// OpenMP compatibility layer
#include <utils/openmp/utils_openmp.hpp>

// T-Rex Selector includes
#include <trex_selector_methods/trex_core/trex.hpp>
#include <utils/datageneration/utils_datagen.hpp>
#include <utils/eval_metrics/utils_eval_cdiagnostics.hpp>
#include <utils/eval_metrics/utils_eval_rates.hpp>

// Demo utilities
#include "trex_sim_utils.hpp"
using namespace trex_sim;

// ==============================================================================
// Namespace aliases
// ==============================================================================

namespace cdianostics = trex::utils::eval::cdiagnostics;
namespace datagen     = trex::utils::datageneration::datagen;
namespace dummygen    = trex::utils::datageneration::dummygen;

// T-Rex types
using trex::trex_selector_methods::trex_core::LLoopStrategy;
using trex::trex_selector_methods::trex_core::TRexControlParameter;
using trex::trex_selector_methods::utils::solver_dispatch::SolverTypeForTRex;

// ==============================================================================


// ==============================================================================
// Shared helpers (self-contained; mirrors demo_trex_06_mmap.cpp)
// ==============================================================================

/**
 * @brief Shared TRexControlParameter baseline for all mmap demos.
 * @note  use_memory_mapping = true couples two behaviours:
 *   (1) Internal dummy matrices D are memory-mapped (TRexMemMapManager).
 *   (2) Solver warm-start state is serialized to disk between T-loop iterations
 *       (SERIALIZED WarmStartMode). Solver footprint scales as O(T * (p + L)).
 */
static TRexControlParameter make_mmap_trex_ctrl() {
    TRexControlParameter ctrl;
    ctrl.K                     = 20;
    ctrl.max_dummy_multiplier  = 10;
    ctrl.use_max_T_stop        = true;
    ctrl.dummy_distribution    = dummygen::Distribution::Normal();
    ctrl.lloop_strategy        = LLoopStrategy::HCONCAT;
    ctrl.tloop_stagnation_stop = true;
    ctrl.use_memory_mapping    = true;  // enables D mmap + solver serialization
    ctrl.solver_type           = SolverTypeForTRex::TLARS;
    return ctrl;
}

// ==============================================================================


// ==============================================================================
// Demo C: Monte Carlo — In-memory X + use_memory_mapping=true
//         Fixed true support; SNR sweep; TLARS solver.
//         Purpose: verify that the D-mmap + solver-serialization pipeline
//         produces stable, reproducible FDR/TPR statistics over many runs.
// ==============================================================================

void demo_TRexSelector_d_mmap_MonteCarlo(std::size_t num_MC, bool high_dim, bool rnd_coef) {

    std::cout.setf(std::ios::unitbuf);

    cdianostics::print_section_header(
        "Demo C: MC — In-Memory X, Solver Serialization + D Memory-Mapping");
    std::cout << (high_dim ? "High-dimensional (p > n)" : "Low-dimensional (n > p)") << "\n";

    const Eigen::Index n = high_dim ? 300 : 1000;
    const Eigen::Index p = high_dim ? 1000 : 300;
    const std::size_t  s = 10;    // cardinality of true support
    const std::vector<double> snr_values = {0.1, 0.5, 1.0, 2.0, 5.0};
    const double tFDR = 0.1;
    std::cout << "n = " << n << ", p = " << p << "\n";

    const std::vector<std::string> solver_names = {"TLARS"};
    std::map<std::string, Eigen::VectorXd> fdr_results_map;
    std::map<std::string, Eigen::VectorXd> tpr_results_map;
    std::map<std::string, Eigen::VectorXd> avg_L_results_map;
    std::map<std::string, Eigen::VectorXd> avg_T_results_map;
    for (const auto& name : solver_names) {
        fdr_results_map[name]   = Eigen::VectorXd(snr_values.size());
        tpr_results_map[name]   = Eigen::VectorXd(snr_values.size());
        avg_L_results_map[name] = Eigen::VectorXd(snr_values.size());
        avg_T_results_map[name] = Eigen::VectorXd(snr_values.size());
    }

    TRexControlParameter trex_ctrl = make_mmap_trex_ctrl();
    trex_ctrl.tloop_max_stagnant_steps = 7;
    trex_ctrl.parallel_rnd_experiments = false;
    trex_ctrl.opt_threshold            = 0.75;

    // ===================================================================
    // Generate fixed true support and coefficients (shared across all MC runs)
    // ===================================================================
    std::mt19937 rng(24);
    std::uniform_int_distribution<std::size_t> udist(0, static_cast<std::size_t>(p) - 1);
    std::normal_distribution<double> ndist(0.0, 1.0);

    std::unordered_set<std::size_t> support_set;
    while (support_set.size() < s) support_set.insert(udist(rng));
    const std::vector<std::size_t> true_support(support_set.begin(), support_set.end());

    std::vector<double> true_coefs;
    true_coefs.reserve(s);
    for (std::size_t i = 0; i < s; ++i)
        true_coefs.push_back(rnd_coef ? ndist(rng) : 1.0);

    std::cout << "True support (cardinality " << s << "): ";
    for (const auto idx : true_support) std::cout << idx << " ";
    std::cout << "\n\n";

    // ===================================================================
    // Solver loop → SNR loop → parallel MC
    // ===================================================================
    for (const auto& solver_name : solver_names) {

        cdianostics::print_section_header("Solver: " + solver_name);

        for (std::size_t snr_idx = 0; snr_idx < snr_values.size(); ++snr_idx) {

            const double snr = snr_values[snr_idx];

            const auto make_data = [=](unsigned seed) -> trex_sim::TrexDGPData {
                datagen::SyntheticData data(
                    n, p, true_support, true_coefs, snr,
                    static_cast<int>(seed));
                return {data.getX(), data.getY(), true_support};
            };

            std::ostringstream lbl;
            lbl << "SNR=" << std::fixed << std::setprecision(2) << snr << " [TLARS/mmap-D]";
            const unsigned base_seed = 24u + static_cast<unsigned>(snr_idx) * 1000u;
            const auto result = trex_sim::run_mc_trials_trex(
                num_MC, lbl.str(), make_data, tFDR, trex_ctrl, base_seed);

            const auto ei = static_cast<Eigen::Index>(snr_idx);
            fdr_results_map[solver_name](ei)   = result.avg_fdr;
            tpr_results_map[solver_name](ei)   = result.avg_tpr;
            avg_L_results_map[solver_name](ei) = result.avg_L;
            avg_T_results_map[solver_name](ei) = result.avg_T;
        }
        std::cout << "\n";
    }

    const std::string stem = "d03_trex_mmap_demo_c_n" + std::to_string(n) +
                             "_p" + std::to_string(p) +
                             "_sw" + std::to_string(trex_ctrl.tloop_max_stagnant_steps);
    save_and_print_mc_results(num_MC, stem, snr_values, solver_names,
                              fdr_results_map, tpr_results_map,
                              avg_L_results_map, avg_T_results_map);
    std::cout << "\n\n";
}


// ==============================================================================
// Demo D: Monte Carlo — Fully memory-mapped X + use_memory_mapping=true
//         Fixed true support; SNR sweep; TLARS solver.
//         Each MC iteration creates and destroys its own mmap-backed X/y files.
//
//         SAFETY NOTE: MmapFileGuard is declared BEFORE SyntheticDataMapped
//         inside the MC scope so that C++ reverse-LIFO destruction gives the
//         correct order: data dtor closes mmap handles → guard dtor removes files.
//         This holds even if TRexSelector::select() throws.
// ==============================================================================

void demo_TRexSelector_full_mmap_MonteCarlo(std::size_t num_MC, bool high_dim, bool rnd_coef) {

    std::cout.setf(std::ios::unitbuf);

    cdianostics::print_section_header(
        "Demo D: MC — Fully Memory-Mapped X + D + Solver Serialization");
    std::cout << (high_dim ? "High-dimensional (p > n)" : "Low-dimensional (n > p)") << "\n";

    const Eigen::Index n = high_dim ? 300 : 1000;
    const Eigen::Index p = high_dim ? 1000 : 300;
    const std::size_t  s = 10;
    const std::vector<double> snr_values = {0.1, 0.5, 1.0, 2.0, 5.0};
    const double tFDR = 0.1;
    std::cout << "n = " << n << ", p = " << p << "\n";

    const std::vector<std::string> solver_names = {"TLARS"};
    std::map<std::string, Eigen::VectorXd> fdr_results_map;
    std::map<std::string, Eigen::VectorXd> tpr_results_map;
    std::map<std::string, Eigen::VectorXd> avg_L_results_map;
    std::map<std::string, Eigen::VectorXd> avg_T_results_map;
    for (const auto& name : solver_names) {
        fdr_results_map[name]   = Eigen::VectorXd(snr_values.size());
        tpr_results_map[name]   = Eigen::VectorXd(snr_values.size());
        avg_L_results_map[name] = Eigen::VectorXd(snr_values.size());
        avg_T_results_map[name] = Eigen::VectorXd(snr_values.size());
    }

    TRexControlParameter trex_ctrl = make_mmap_trex_ctrl();
    trex_ctrl.tloop_max_stagnant_steps = 7;
    trex_ctrl.parallel_rnd_experiments = false;
    trex_ctrl.opt_threshold            = 0.75;

    // ===================================================================
    // Generate fixed true support and coefficients (shared across all MC runs)
    // ===================================================================
    std::mt19937 rng(24);
    std::uniform_int_distribution<std::size_t> udist(0, static_cast<std::size_t>(p) - 1);
    std::normal_distribution<double> ndist(0.0, 1.0);

    std::unordered_set<std::size_t> support_set;
    while (support_set.size() < s) support_set.insert(udist(rng));
    const std::vector<std::size_t> true_support(support_set.begin(),
                                                support_set.end());

    std::vector<double> true_coefs;
    true_coefs.reserve(s);
    for (std::size_t i = 0; i < s; ++i)
        true_coefs.push_back(rnd_coef ? ndist(rng) : 1.0);

    std::cout << "True support (cardinality " << s << "): ";
    for (const auto idx : true_support) std::cout << idx << " ";
    std::cout << "\n\n";

    // ===================================================================
    // Solver loop → SNR loop → parallel MC (per-thread mmap backing files)
    // ===================================================================
    // MmapFileGuard is declared BEFORE SyntheticDataMapped inside the factory
    // lambda body (LIFO: mmap handles closed before files removed, exception-safe).
    // Each thread uses unique file paths via omp_get_thread_num().
    for (const auto& solver_name : solver_names) {

        cdianostics::print_section_header("Solver: " + solver_name);

        for (std::size_t snr_idx = 0; snr_idx < snr_values.size(); ++snr_idx) {

            const double snr = snr_values[snr_idx];

            const auto make_data_mmap = [=](unsigned seed,
                                            const std::string& X_path,
                                            const std::string& y_path) -> trex_sim::TrexDGPData {
                // Guard declared BEFORE SyntheticDataMapped (LIFO cleanup)
                MmapFileGuard guard{{X_path, y_path}};
                datagen::SyntheticDataMapped sdm(
                    X_path, y_path,
                    n, p, true_support, true_coefs, snr,
                    static_cast<int>(seed));
                Eigen::MatrixXd X_copy = sdm.getX();
                Eigen::VectorXd y_copy = Eigen::Map<Eigen::VectorXd>(
                    sdm.getY().data(), sdm.rows());
                return {std::move(X_copy), std::move(y_copy), true_support};
            };

            std::ostringstream lbl;
            lbl << "SNR=" << std::fixed << std::setprecision(2) << snr << " [TLARS/mmap-XD]";
            const unsigned base_seed = 24u + static_cast<unsigned>(snr_idx) * 1000u;
            const auto result = trex_sim::run_mc_trials_trex_mmap(
                num_MC,
                lbl.str(),
                make_data_mmap,
                tFDR,
                trex_ctrl,
                base_seed,
                "X_mmap_mc"
            );

            const auto ei = static_cast<Eigen::Index>(snr_idx);
            fdr_results_map[solver_name](ei)   = result.avg_fdr;
            tpr_results_map[solver_name](ei)   = result.avg_tpr;
            avg_L_results_map[solver_name](ei) = result.avg_L;
            avg_T_results_map[solver_name](ei) = result.avg_T;
        }
        std::cout << "\n";
    }

    const std::string stem = "d03_trex_mmap_demo_d_n" + std::to_string(n) +
                             "_p" + std::to_string(p) +
                             "_sw" + std::to_string(trex_ctrl.tloop_max_stagnant_steps);
    save_and_print_mc_results(num_MC, stem, snr_values, solver_names,
                              fdr_results_map, tpr_results_map,
                              avg_L_results_map, avg_T_results_map);
    std::cout << "\n\n";
}



// ==============================================================================
// Main
// ==============================================================================

int main() {

    std::cout.setf(std::ios::unitbuf);
    omp_set_num_threads(6);
    std::cout << "Running with " << omp_get_max_threads() << " threads\n\n";

    // ============================================================
    // Demo C: MC — in-memory X + solver serialization + D mmap
    // ============================================================
    if (true)
        demo_TRexSelector_d_mmap_MonteCarlo(
            10,
            true,
            false
        );

    // ============================================================
    // Demo D: MC — fully memory-mapped pipeline (X + D + solver serialization)
    // ============================================================
    if (true)
        demo_TRexSelector_full_mmap_MonteCarlo(
            500,
            true,
            false
        );

    return 0;
}
